File: lib/data_frame.ml
open! Core

type t = Data_frame0.t

external create : Series.t list -> (t, string) result = "rust_data_frame_new"

let create_exn series = create series |> Utils.string_result_ok_exn

external read_csv
  :  string
  -> schema:Schema.t option
  -> try_parse_dates:bool option
  -> (t, string) result
  = "rust_data_frame_read_csv"

let read_csv ?schema ?try_parse_dates path = read_csv path ~schema ~try_parse_dates

let read_csv_exn ?schema ?try_parse_dates path =
  read_csv ?schema ?try_parse_dates path |> Utils.string_result_ok_exn
;;

external write_csv : t -> string -> (unit, string) result = "rust_data_frame_write_csv"

let write_csv_exn t path = write_csv t path |> Utils.string_result_ok_exn

external read_parquet : string -> (t, string) result = "rust_data_frame_read_parquet"

let read_parquet_exn path = read_parquet path |> Utils.string_result_ok_exn

external write_parquet
  :  t
  -> string
  -> (unit, string) result
  = "rust_data_frame_write_parquet"

let write_parquet_exn t path = write_parquet t path |> Utils.string_result_ok_exn

external read_json : string -> (t, string) result = "rust_data_frame_read_json"

let read_json_exn path = read_json path |> Utils.string_result_ok_exn

external write_json : t -> string -> (unit, string) result = "rust_data_frame_write_json"

let write_json_exn t path = write_json t path |> Utils.string_result_ok_exn

external read_jsonl : string -> (t, string) result = "rust_data_frame_read_jsonl"

let read_jsonl_exn path = read_jsonl path |> Utils.string_result_ok_exn

external write_jsonl
  :  t
  -> string
  -> (unit, string) result
  = "rust_data_frame_write_jsonl"

let write_jsonl_exn t path = write_jsonl t path |> Utils.string_result_ok_exn

external clear : t -> t = "rust_data_frame_clear"

external describe
  :  t
  -> percentiles:float list option
  -> (t, string) result
  = "rust_data_frame_describe"

let describe ?percentiles t = describe t ~percentiles
let describe_exn ?percentiles t = describe ?percentiles t |> Utils.string_result_ok_exn

external height : t -> int = "rust_data_frame_height"
external estimated_size : t -> int = "rust_data_frame_estimated_size"
external lazy_ : t -> Lazy_frame.t = "rust_data_frame_lazy"

let in_lazy t ~f = lazy_ t |> f |> Lazy_frame.collect
let in_lazy_exn t ~f = lazy_ t |> f |> Lazy_frame.collect_exn
let select t ~exprs = in_lazy t ~f:(Lazy_frame.select ~exprs)
let select_exn t ~exprs = in_lazy_exn t ~f:(Lazy_frame.select ~exprs)
let with_columns t ~exprs = in_lazy t ~f:(Lazy_frame.with_columns ~exprs)
let with_columns_exn t ~exprs = in_lazy_exn t ~f:(Lazy_frame.with_columns ~exprs)
let groupby ?is_stable t ~by ~agg = in_lazy t ~f:(Lazy_frame.groupby ?is_stable ~by ~agg)

let groupby_exn ?is_stable t ~by ~agg =
  in_lazy_exn t ~f:(Lazy_frame.groupby ?is_stable ~by ~agg)
;;

let groupby_dynamic
  ?every
  ?period
  ?offset
  ?truncate
  ?include_boundaries
  ?closed_window
  ?start_by
  ?check_sorted
  t
  ~index_column
  ~by
  ~agg
  =
  in_lazy
    t
    ~f:
      (Lazy_frame.groupby_dynamic
         ?every
         ?period
         ?offset
         ?truncate
         ?include_boundaries
         ?closed_window
         ?start_by
         ?check_sorted
         ~index_column
         ~by
         ~agg)
;;

let groupby_dynamic_exn
  ?every
  ?period
  ?offset
  ?truncate
  ?include_boundaries
  ?closed_window
  ?start_by
  ?check_sorted
  t
  ~index_column
  ~by
  ~agg
  =
  in_lazy_exn
    t
    ~f:
      (Lazy_frame.groupby_dynamic
         ?every
         ?period
         ?offset
         ?truncate
         ?include_boundaries
         ?closed_window
         ?start_by
         ?check_sorted
         ~index_column
         ~by
         ~agg)
;;

external column : t -> name:string -> (Series.t, string) result = "rust_data_frame_column"

let column_exn t ~name = column t ~name |> Utils.string_result_ok_exn

external columns
  :  t
  -> names:string list
  -> (Series.t list, string) result
  = "rust_data_frame_columns"

let columns_exn t ~names = columns t ~names |> Utils.string_result_ok_exn

external get_column_names : t -> string list = "rust_data_frame_get_column_names"

external vertical_concat
  :  t list
  -> (t, string) result
  = "rust_data_frame_vertical_concat"

external horizontal_concat
  :  t list
  -> (t, string) result
  = "rust_data_frame_horizontal_concat"

external diagonal_concat
  :  t list
  -> (t, string) result
  = "rust_data_frame_diagonal_concat"

let concat ?(how = `Vertical) ts =
  match how with
  | `Vertical -> vertical_concat ts
  | `Horizontal -> horizontal_concat ts
  | `Diagonal -> diagonal_concat ts
;;

let concat_exn ?how ts = concat ?how ts |> Utils.string_result_ok_exn

external as_single_chunk_par : t -> unit = "rust_data_frame_as_single_chunk_par"
external vstack : t -> other:t -> (unit, string) result = "rust_data_frame_vstack"

let vstack_exn t ~other = vstack t ~other |> Utils.string_result_ok_exn

external pivot
  :  t
  -> values:string list
  -> index:string list
  -> columns:string list
  -> sort_columns:bool
  -> agg_expr:Expr.t option
  -> separator:string option
  -> stable:bool
  -> (t, string) result
  = "rust_data_frame_pivot_bytecode" "rust_data_frame_pivot"

let pivot
  ?agg_expr
  ?(sort_columns = false)
  ?separator
  ?(stable = true)
  t
  ~values
  ~index
  ~columns
  =
  let agg_expr =
    Option.map
      agg_expr
      ~f:
        Expr.(
          function
          | `First -> element () |> first
          | `Sum -> element () |> sum
          | `Max -> element () |> max
          | `Min -> element () |> min
          | `Mean -> element () |> mean
          | `Median -> element () |> median
          | `Last -> element () |> last
          | `Count -> count_ ()
          | `Expr expr -> expr)
  in
  pivot t ~values ~index ~columns ~sort_columns ~agg_expr ~separator ~stable
;;

let pivot_exn ?agg_expr ?sort_columns ?separator ?stable t ~values ~index ~columns =
  pivot ?agg_expr ?sort_columns ?separator ?stable t ~values ~index ~columns
  |> Utils.string_result_ok_exn
;;

external melt
  :  t
  -> id_vars:string list
  -> value_vars:string list
  -> variable_name:string option
  -> value_name:string option
  -> streamable:bool
  -> (t, string) result
  = "rust_data_frame_melt_bytecode" "rust_data_frame_melt"

let melt ?variable_name ?value_name ?(streamable = false) t ~id_vars ~value_vars =
  melt t ~id_vars ~value_vars ~variable_name ~value_name ~streamable
;;

let melt_exn ?variable_name ?value_name ?streamable t ~id_vars ~value_vars =
  melt ?variable_name ?value_name ?streamable t ~id_vars ~value_vars
  |> Utils.string_result_ok_exn
;;

external sort
  :  t
  -> by_column:string list
  -> descending:bool list
  -> maintain_order:bool
  -> (t, string) result
  = "rust_data_frame_sort"

let sort ?descending ?(maintain_order = true) t ~by_column =
  let descending =
    Option.value descending ~default:(List.map by_column ~f:(Fn.const false))
  in
  sort t ~by_column ~descending ~maintain_order
;;

let sort_exn ?descending ?maintain_order t ~by_column =
  sort ?descending ?maintain_order t ~by_column |> Utils.string_result_ok_exn
;;

external head : t -> length:int option -> t = "rust_data_frame_head"

let head ?length t = head t ~length

external tail : t -> length:int option -> t = "rust_data_frame_tail"

let tail ?length t = tail t ~length

external sample_n
  :  t
  -> n:int
  -> with_replacement:bool
  -> shuffle:bool
  -> seed:int option
  -> (t, string) result
  = "rust_data_frame_sample_n"

let sample_n ?seed t ~n ~with_replacement ~shuffle =
  sample_n t ~n ~with_replacement ~shuffle ~seed
;;

let sample_n_exn ?seed t ~n ~with_replacement ~shuffle =
  sample_n ?seed t ~n ~with_replacement ~shuffle |> Utils.string_result_ok_exn
;;

external sum : t -> t = "rust_data_frame_sum"
external mean : t -> t = "rust_data_frame_mean"
external median : t -> t = "rust_data_frame_median"
external null_count : t -> t = "rust_data_frame_null_count"

external fill_null
  :  t
  -> strategy:Fill_null_strategy.t
  -> (t, string) result
  = "rust_data_frame_fill_null_with_strategy"

let fill_null_exn t ~strategy = fill_null t ~strategy |> Utils.string_result_ok_exn

external interpolate
  :  t
  -> method_:[ `Linear | `Nearest ]
  -> (t, string) result
  = "rust_data_frame_interpolate"

let interpolate_exn t ~method_ = interpolate t ~method_ |> Utils.string_result_ok_exn

external upsample
  :  t
  -> by:string list
  -> time_column:string
  -> every:string
  -> offset:string
  -> stable:bool
  -> (t, string) result
  = "rust_data_frame_upsample_bytecode" "rust_data_frame_upsample"

let upsample ?(stable = true) t ~by ~time_column ~every ~offset =
  upsample t ~by ~time_column ~every ~offset ~stable
;;

let upsample_exn ?stable t ~by ~time_column ~every ~offset =
  upsample ?stable t ~by ~time_column ~every ~offset |> Utils.string_result_ok_exn
;;

external explode
  :  t
  -> columns:string list
  -> (t, string) result
  = "rust_data_frame_explode"

let explode_exn t ~columns = explode t ~columns |> Utils.string_result_ok_exn

external schema : t -> Schema.t = "rust_data_frame_schema"
external to_string_hum : t -> string = "rust_data_frame_to_string_hum"

external partition_by'
  :  t
  -> by:string list
  -> maintain_order:bool
  -> (t list, string) result
  = "rust_data_frame_partition_by"

let partition_by ?(maintain_order = true) t ~by = partition_by' t ~by ~maintain_order

let partition_by_exn ?maintain_order t ~by =
  partition_by ?maintain_order t ~by |> Utils.string_result_ok_exn
;;

let print t = print_endline (to_string_hum t)

include Pretty_printer.Register (struct
    type nonrec t = t

    let module_name = "Polars.Data_frame"
    let to_string = to_string_hum
  end)

module Expert = struct
  external modify_series_at_chunk_index_exn
    :  t
    -> dtype:'a Data_type.Typed.t
    -> series_index:int
    -> chunk_index:int
    -> indices_and_values:(int * 'a) list
    -> (unit, string) result
    = "rust_data_frame_modify_series_at_chunk_index"

  let modify_series_at_chunk_index t ~dtype ~series_index ~chunk_index ~indices_and_values
    =
    match
      Or_error.try_with (fun () ->
        modify_series_at_chunk_index_exn
          t
          ~dtype
          ~series_index
          ~chunk_index
          ~indices_and_values)
    with
    | Ok result -> result
    | Error error ->
      Error
        (sprintf
           "modify_series_at_chunk_index_exn raised an exception. Usually this happens \
            when accessing an index out of bounds of the chunk or passing in a value \
            outside of the domain of dtype: %s"
           (Error.to_string_hum error))
  ;;

  external modify_optional_series_at_chunk_index_exn
    :  t
    -> dtype:'a Data_type.Typed.t
    -> series_index:int
    -> chunk_index:int
    -> indices_and_values:(int * 'a option) list
    -> (unit, string) result
    = "rust_data_frame_modify_optional_series_at_chunk_index"

  let modify_optional_series_at_chunk_index
    t
    ~dtype
    ~series_index
    ~chunk_index
    ~indices_and_values
    =
    match
      Or_error.try_with (fun () ->
        modify_optional_series_at_chunk_index_exn
          t
          ~dtype
          ~series_index
          ~chunk_index
          ~indices_and_values)
    with
    | Ok result -> result
    | Error error ->
      Error
        (sprintf
           "modify_series_at_chunk_index_exn raised an exception. Usually this happens \
            when accessing an index out of bounds of the chunk or passing in a value \
            outside of the domain of dtype: %s"
           (Error.to_string_hum error))
  ;;

  external clear_mut : t -> unit = "rust_data_frame_clear_mut"
end

File: rust/src/data_frame.rs
use crate::utils::*;
use ocaml_interop::{
    ocaml_export, DynBox, OCaml, OCamlFloat, OCamlInt, OCamlList, OCamlRef, ToOCaml,
};
use polars::prelude::*;
use smartstring::{LazyCompact, SmartString};
use std::fs::File;

ocaml_export! {
    fn rust_data_frame_new(cr, series: OCamlRef<OCamlList<DynBox<Series>>>) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let series: Vec<Series> = unwrap_abstract_vec(series.to_rust(cr));

        DataFrame::new(series).map(Abstract).map_err(|err| err.to_string()).to_ocaml(cr)
    }

    fn rust_data_frame_read_csv(
        cr,
        path: OCamlRef<String>,
        schema: OCamlRef<Option<DynBox<Schema>>>,
        try_parse_dates: OCamlRef<Option<bool>>
    ) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let path: String = path.to_rust(cr);
        let schema = schema.to_rust::<Option<Abstract<Schema>>>(cr).map(|Abstract(schema)| Arc::new(schema));
        let try_parse_dates: Option<bool> = try_parse_dates.to_rust(cr);

        CsvReader::from_path(&path)
        .and_then(|csv_reader| {
            let csv_reader = csv_reader.with_dtypes(schema);
            match try_parse_dates {
                None => csv_reader,
                Some(try_parse_dates) => csv_reader.with_try_parse_dates(try_parse_dates),
            }.finish()
        })
        .map(Abstract).map_err(|err| err.to_string()).to_ocaml(cr)
    }

    fn rust_data_frame_write_csv(
        cr,
        data_frame: OCamlRef<DynBox<DataFrame>>,
        path: OCamlRef<String>,
    ) -> OCaml<Result<(),String>> {
        let Abstract(mut data_frame) = data_frame.to_rust(cr);
        let path: String = path.to_rust(cr);

        File::create(&path).map_err(|err| err.to_string())
        .and_then(|file|
            CsvWriter::new(&file)
            .finish(&mut data_frame)
            .map_err(|err| err.to_string())
        )
        .to_ocaml(cr)
    }

    fn rust_data_frame_read_parquet(
        cr,
        path: OCamlRef<String>,
    ) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let path: String = path.to_rust(cr);

        File::open(&path).map_err(|err| err.to_string())
        .and_then(|file|
            ParquetReader::new(file).finish().map_err(|err| err.to_string()))
        .map(Abstract).to_ocaml(cr)
    }

    fn rust_data_frame_write_parquet(
        cr,
        data_frame: OCamlRef<DynBox<DataFrame>>,
        path: OCamlRef<String>,
    ) -> OCaml<Result<(),String>> {
        let Abstract(mut data_frame) = data_frame.to_rust(cr);
        let path: String = path.to_rust(cr);

        File::create(&path).map_err(|err| err.to_string())
        .and_then(|file|
            ParquetWriter::new(file).finish(&mut data_frame)
            .map(|_file_size_in_bytes| ()).map_err(|err| err.to_string()))
        .to_ocaml(cr)
    }

    fn rust_data_frame_read_json(
        cr,
        path: OCamlRef<String>,
    ) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let path: String = path.to_rust(cr);

        File::open(&path).map_err(|err| err.to_string())
        .and_then(|file|
            JsonReader::new(file).finish().map_err(|err| err.to_string()))
        .map(Abstract).to_ocaml(cr)
    }

    fn rust_data_frame_write_json(
        cr,
        data_frame: OCamlRef<DynBox<DataFrame>>,
        path: OCamlRef<String>,
    ) -> OCaml<Result<(),String>> {
        let Abstract(mut data_frame) = data_frame.to_rust(cr);
        let path: String = path.to_rust(cr);

        File::create(&path).map_err(|err| err.to_string())
        .and_then(|file|
            JsonWriter::new(file).with_json_format(JsonFormat::Json).finish(&mut data_frame)
            .map_err(|err| err.to_string()))
        .to_ocaml(cr)
    }

    fn rust_data_frame_read_jsonl(
        cr,
        path: OCamlRef<String>,
    ) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let path: String = path.to_rust(cr);

        File::open(&path).map_err(|err| err.to_string())
        .and_then(|file|
            JsonLineReader::new(file).finish().map_err(|err| err.to_string()))
        .map(Abstract).to_ocaml(cr)
    }

    fn rust_data_frame_write_jsonl(
        cr,
        data_frame: OCamlRef<DynBox<DataFrame>>,
        path: OCamlRef<String>,
    ) -> OCaml<Result<(),String>> {
        let Abstract(mut data_frame) = data_frame.to_rust(cr);
        let path: String = path.to_rust(cr);

        File::create(&path).map_err(|err| err.to_string())
        .and_then(|file|
            JsonWriter::new(file).with_json_format(JsonFormat::JsonLines).finish(&mut data_frame)
            .map_err(|err| err.to_string()))
        .to_ocaml(cr)
    }

    fn rust_data_frame_describe(cr, data_frame: OCamlRef<DynBox<DataFrame>>, percentiles: OCamlRef<Option<OCamlList<OCamlFloat>>>) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        let percentiles: Option<Vec<f64>> = percentiles.to_rust(cr);

        // TODO: I'm not sure why I can't do this with something like
        // .map(|percentiles| percentiles.as_slice()
        match percentiles {
            None => data_frame.describe(None),
            Some(percentiles) => data_frame.describe(Some(percentiles.as_slice()))
        }
        .map(Abstract).map_err(|err| err.to_string()).to_ocaml(cr)
    }

   fn rust_data_frame_lazy(cr, data_frame: OCamlRef<DynBox<DataFrame>>) -> OCaml<DynBox<LazyFrame>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        OCaml::box_value(cr, data_frame.lazy())
    }

    fn rust_data_frame_column(cr, data_frame: OCamlRef<DynBox<DataFrame>>, name: OCamlRef<String>) -> OCaml<Result<DynBox<Series>,String>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        let name: String = name.to_rust(cr);
        data_frame.column(&name)
        .map(|series| Abstract(series.clone()))
        .map_err(|err| err.to_string())
        .to_ocaml(cr)
    }

    fn rust_data_frame_columns(cr, data_frame: OCamlRef<DynBox<DataFrame>>, names: OCamlRef<OCamlList<String>>) -> OCaml<Result<OCamlList<DynBox<Series>>,String>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        let names: Vec<String> = names.to_rust(cr);
        data_frame.columns(&names)
        .map(|series|
            series
            .into_iter()
            .map(|series| Abstract(series.clone()))
            .collect::<Vec<Abstract<_>>>())
        .map_err(|err| err.to_string())
        .to_ocaml(cr)
    }

    fn rust_data_frame_get_column_names(cr, data_frame: OCamlRef<DynBox<DataFrame>>) -> OCaml<OCamlList<String>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        data_frame.get_column_names().to_ocaml(cr)
    }

    fn rust_data_frame_vertical_concat(cr, data_frames: OCamlRef<OCamlList<DynBox<DataFrame>>>) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let data_frames = unwrap_abstract_vec(data_frames.to_rust(cr));

        let result: PolarsResult<_> = try {
            let mut data_frames = data_frames.into_iter();
            let first = data_frames.next().ok_or(PolarsError::NoData("No dataframes provided for vertical concatenation".into()))?;
            let mut result = first.clone();
            for data_frame in data_frames {
                result = result.vstack(&data_frame)?;
            }
            result
        };
        result.map(Abstract).map_err(|err| err.to_string()).to_ocaml(cr)
    }

    fn rust_data_frame_horizontal_concat(cr, data_frames: OCamlRef<OCamlList<DynBox<DataFrame>>>) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let data_frames = unwrap_abstract_vec(data_frames.to_rust(cr));

        polars::functions::hor_concat_df(&data_frames).map(Abstract).map_err(|err| err.to_string()).to_ocaml(cr)
    }

    fn rust_data_frame_diagonal_concat(cr, data_frames: OCamlRef<OCamlList<DynBox<DataFrame>>>) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let data_frames = unwrap_abstract_vec(data_frames.to_rust(cr));

        polars::functions::diag_concat_df(&data_frames).map(Abstract).map_err(|err| err.to_string()).to_ocaml(cr)
    }

    fn rust_data_frame_pivot|rust_data_frame_pivot_bytecode(cr,
        data_frame: OCamlRef<DynBox<DataFrame>>,
        values: OCamlRef<OCamlList<String>>,
        index: OCamlRef<OCamlList<String>>,
        columns: OCamlRef<OCamlList<String>>,
        sort_columns: OCamlRef<bool>,
        agg_expr: OCamlRef<Option<DynBox<Expr>>>,
        separator: OCamlRef<Option<String>>,
        stable: OCamlRef<bool>
    )
    -> OCaml<Result<DynBox<DataFrame>,String>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);

        let values: Vec<String> = values.to_rust(cr);
        let index: Vec<String> = index.to_rust(cr);
        let columns: Vec<String> = columns.to_rust(cr);
        let sort_columns: bool = sort_columns.to_rust(cr);
        let agg_expr: Option<Expr> = agg_expr.to_rust::<Option<Abstract<Expr>>>(cr).map(|Abstract(expr)| expr);
        let separator: Option<String> = separator.to_rust(cr);

        let stable: bool = stable.to_rust(cr);

        if stable {
            pivot::pivot_stable(&data_frame, &values, &index, &columns, sort_columns, agg_expr, separator.as_deref())
        } else {
            pivot::pivot(&data_frame, &values, &index, &columns, sort_columns, agg_expr, separator.as_deref())
        }.map(Abstract).map_err(|err| err.to_string()).to_ocaml(cr)
    }

    fn rust_data_frame_melt|rust_data_frame_melt_bytecode(cr,
        data_frame: OCamlRef<DynBox<DataFrame>>,
        id_vars: OCamlRef<OCamlList<String>>,
        value_vars: OCamlRef<OCamlList<String>>,
        variable_name: OCamlRef<Option<String>>,
        value_name: OCamlRef<Option<String>>,
        streamable: OCamlRef<bool>,
    ) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);

        let id_vars: Vec<SmartString<LazyCompact>> = id_vars.to_rust::<Vec<String>>(cr).into_iter().map(|s| s.into()).collect();
        let value_vars: Vec<SmartString<LazyCompact>> = value_vars.to_rust::<Vec<String>>(cr).into_iter().map(|s| s.into()).collect();
        let variable_name: Option<SmartString<LazyCompact>> = variable_name.to_rust::<Option<String>>(cr).map(|s| s.into());
        let value_name: Option<SmartString<LazyCompact>> = value_name.to_rust::<Option<String>>(cr).map(|s| s.into());
        let streamable: bool = streamable.to_rust(cr);

        let melt_args = MeltArgs {
            id_vars, value_vars, variable_name, value_name, streamable
        };
        data_frame.melt2(melt_args).map(Abstract).map_err(|err| err.to_string()).to_ocaml(cr)
    }

    fn rust_data_frame_sort(cr, data_frame: OCamlRef<DynBox<DataFrame>>, by_column: OCamlRef<OCamlList<String>>, descending: OCamlRef<OCamlList<bool>>, maintain_order: OCamlRef<bool>) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        let by_column: Vec<String> = by_column.to_rust(cr);
        let descending: Vec<bool> = descending.to_rust(cr);
        let maintain_order: bool = maintain_order.to_rust(cr);

        data_frame.sort(&by_column, descending, maintain_order)
        .map(Abstract).map_err(|err| err.to_string()).to_ocaml(cr)
    }

    fn rust_data_frame_head(
        cr,
        data_frame: OCamlRef<DynBox<DataFrame>>,
        length: OCamlRef<Option<OCamlInt>>
    ) -> OCaml<DynBox<DataFrame>> {

        let Abstract(data_frame) = data_frame.to_rust(cr);
        let length = length.to_rust::<Coerce<_, Option<i64>, Option<usize>>>(cr).get();

        Abstract(data_frame.head(length)).to_ocaml(cr)
    }

    fn rust_data_frame_tail(
        cr,
        data_frame: OCamlRef<DynBox<DataFrame>>,
        length: OCamlRef<Option<OCamlInt>>
    ) -> OCaml<DynBox<DataFrame>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        let length = length.to_rust::<Coerce<_, Option<i64>, Option<usize>>>(cr).get();

        Abstract(data_frame.tail(length)).to_ocaml(cr)
    }

    fn rust_data_frame_sample_n(
        cr,
        data_frame: OCamlRef<DynBox<DataFrame>>,
        n: OCamlRef<OCamlInt>,
        with_replacement: OCamlRef<bool>,
        shuffle: OCamlRef<bool>,
        seed: OCamlRef<Option<OCamlInt>>
    ) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        let n = n.to_rust::<Coerce<_, i64, usize>>(cr).get();
        let with_replacement: bool = with_replacement.to_rust(cr);
        let shuffle: bool = shuffle.to_rust(cr);
        let seed: Option<u64> = seed.to_rust::<Coerce<_, Option<i64>, Option<u64>>>(cr).get();

        data_frame.sample_n(n, with_replacement, shuffle, seed)
        .map(Abstract).map_err(|err| err.to_string())
        .to_ocaml(cr)
    }

    fn rust_data_frame_sum(cr, data_frame: OCamlRef<DynBox<DataFrame>>) -> OCaml<DynBox<DataFrame>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        Abstract(data_frame.sum()).to_ocaml(cr)
    }

    fn rust_data_frame_mean(cr, data_frame: OCamlRef<DynBox<DataFrame>>) -> OCaml<DynBox<DataFrame>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        Abstract(data_frame.mean()).to_ocaml(cr)
    }

    fn rust_data_frame_median(cr, data_frame: OCamlRef<DynBox<DataFrame>>) -> OCaml<DynBox<DataFrame>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        Abstract(data_frame.median()).to_ocaml(cr)
    }

    fn rust_data_frame_null_count(cr, data_frame: OCamlRef<DynBox<DataFrame>>) -> OCaml<DynBox<DataFrame>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        Abstract(data_frame.null_count()).to_ocaml(cr)
    }

    fn rust_data_frame_fill_null_with_strategy(
        cr,
        data_frame: OCamlRef<DynBox<DataFrame>>,
        strategy: OCamlRef<FillNullStrategy>
    ) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        let PolarsFillNullStrategy(strategy) = strategy.to_rust(cr);

        data_frame.fill_null(strategy)
        .map(Abstract).map_err(|err| err.to_string())
        .to_ocaml(cr)
    }

    fn rust_data_frame_interpolate(
        cr,
        data_frame: OCamlRef<DynBox<DataFrame>>,
        method: OCamlRef<InterpolationMethod>,
    ) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        let PolarsInterpolationMethod(method) = method.to_rust(cr);

        let series =
            data_frame.get_columns()
            .into_iter()
            .map(|series| interpolate(series, method))
            .collect::<Vec<_>>();

        DataFrame::new(series)
        .map(Abstract)
        .map_err(|err| err.to_string())
        .to_ocaml(cr)
    }

    fn rust_data_frame_upsample|rust_data_frame_upsample_bytecode(
        cr,
        data_frame: OCamlRef<DynBox<DataFrame>>,
        by: OCamlRef<OCamlList<String>>,
        time_column: OCamlRef<String>,
        every: OCamlRef<String>,
        offset: OCamlRef<String>,
        stable: OCamlRef<bool>,
    ) -> OCaml<Result<DynBox<DataFrame>,String>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        let by: Vec<String> = by.to_rust(cr);
        let time_column: String = time_column.to_rust(cr);
        let every: String = every.to_rust(cr);
        let offset: String = offset.to_rust(cr);
        let stable: bool = stable.to_rust(cr);

        if stable {
            data_frame.upsample_stable(&by, &time_column, Duration::parse(&every), Duration::parse(&offset))
        } else {
            data_frame.upsample(&by, &time_column, Duration::parse(&every), Duration::parse(&offset))
        }.map(Abstract).map_err(|err| err.to_string()).to_ocaml(cr)
    }

    fn rust_data_frame_explode(cr, data_frame: OCamlRef<DynBox<DataFrame>>, columns: OCamlRef<OCamlList<String>>) -> OCaml<Result<DynBox<DataFrame>, String>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        let columns: Vec<String> = columns.to_rust(cr);

        data_frame.explode(columns)
        .map(Abstract)
        .map_err(|err| err.to_string())
        .to_ocaml(cr)
    }

    fn rust_data_frame_schema(cr, data_frame: OCamlRef<DynBox<DataFrame>>) -> OCaml<DynBox<Schema>> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        OCaml::box_value(cr, data_frame.schema())
    }

    fn rust_data_frame_to_string_hum(cr, data_frame: OCamlRef<DynBox<DataFrame>>) -> OCaml<String> {
        let Abstract(data_frame) = data_frame.to_rust(cr);
        data_frame.to_string().to_ocaml(cr)
    }
}
